File: ios/Anemone SDK/model/config/Sport.swift
//
//  Sport.swift
//  Anemone SDK
//
//  Created by Manuel Gonzalez Villegas on 9/12/16.
//  Copyright © 2016 Syltek Solutions S.L. All rights reserved.
//

import Foundation

public struct SportId: CustomStringConvertible, Hashable {
    public let value: String

    public init(_ value: Any) throws {
        if let value = value as? Int {
            self.init(value)
        } else if let value = value as? String {
            self.init(value)
        } else {
            throw AnemoneException.jsonInvalidFormat(key: nil)
        }
    }

    public init(_ value: Int) {
        self.init("\(value)")
    }

    public init(_ value: String) {
        self.value = value == "1" ? "PADEL"
            : value == "2" ? "TENNIS"
            : value
    }

    public var description: String {
        value
    }

    // swiftlint:disable identifier_name
    public static let PADEL = SportId("PADEL")
    public static let TENNIS = SportId("TENNIS")
    public static let PADBOL = SportId("PADBOL")
    public static let SQUASH = SportId("SQUASH")
    public static let BADMINTON = SportId("BADMINTON")
    public static let FOOTBALL7 = SportId("FOOTBALL7")
    public static let FOOTBALL11 = SportId("FOOTBALL11")
    public static let FUTSAL = SportId("FUTSAL")
    public static let FOOTBALL_OTHERS = SportId("FOOTBALL_OTHERS")
    public static let PICKLEBALL = SportId("PICKLEBALL")
    public static let BEACH_TENNIS = SportId("BEACH_TENNIS")
    public static let BEACH_VOLLEY = SportId("BEACH_VOLLEY")
    public static let BASKETBALL = SportId("BASKETBALL")
    public static let VOLLEYBALL = SportId("VOLLEYBALL")
    public static let TABLE_TENNIS = SportId("TABLE_TENNIS")
    public static let CRICKET = SportId("CRICKET")
    // swiftlint:enable identifier_name
}

public struct Sport {
    public let id: SportId
    public let name: String
    public let isRacket: Bool
    public let properties: [Property]
    public let warnings: [SportWarning]
}

extension Sport: JSONMappable {
    public init(json: JSONObject) throws {
        self.id = try SportId(json.getAny("sport_id"))
        self.name = try json.getString("name")
        self.isRacket = try json.getBoolean("is_racket")
        self.properties = try json.getJSONArray("properties").flatMap { (json: JSONObject) in
            try? Property(json: json)
        }
        self.warnings = (try? json.getJSONArray("warnings").flatMap { (json: JSONObject) in
            try? SportWarning(json: json)
        }) ?? []
    }
}

public extension Sport {
    func propertyWithId(_ id: PropertyId) -> Property? {
        properties.filter { $0.id == id }.first
    }

    func warningFor(propertyId: PropertyId, optionId: PropertyOptionId) -> SportWarning? {
        warnings.filter { $0.propertyId == propertyId && $0.optionId == optionId }.first
    }

    func filterProperties(_ filterIds: [PropertyId]) -> [Property] {
        properties.filter {
            filterIds.map { $0.value }.contains($0.id.value)
        }
    }
}

extension Sport: Equatable { }
public func == (lhs: Sport, rhs: Sport) -> Bool {
    lhs.id == rhs.id
}

File: android/app/src/main/java/com/playtomic/test/anemone_sdk/model/config/Sport.kt
package com.anemonesdk.model.config

import com.anemonesdk.general.json.JSONMappable
import com.anemonesdk.general.json.JSONObject
import com.anemonesdk.model.generic.Id
import org.json.JSONException

/**
 * Created by agarcia on 23/12/2016.
 */

open class SportId : Id {

    @Throws(JSONException::class)
    constructor(value: Any) :
        super(
            value.toString().let { stringValue ->
                if (stringValue == "1") "PADEL"
                else if (stringValue == "2") "TENNIS"
                else stringValue
            }
        )

    companion object {
        val PADEL = SportId("PADEL")
        val TENNIS = SportId("TENNIS")
        val PADBOL = SportId("PADBOL")
        val SQUASH = SportId("SQUASH")
        val BADMINTON = SportId("BADMINTON")
        val FOOTBALL7 = SportId("FOOTBALL7")
        val FOOTBALL11 = SportId("FOOTBALL11")
        val FUTSAL = SportId("FUTSAL")
        val FOOTBALL_OTHERS = SportId("FOOTBALL_OTHERS")
        val PICKLEBALL = SportId("PICKLEBALL")
        val BEACH_TENNIS = SportId("BEACH_TENNIS")
        val BEACH_VOLLEY = SportId("BEACH_VOLLEY")
        val BASKETBALL = SportId("BASKETBALL")
        val VOLLEYBALL = SportId("VOLLEYBALL")
        val TABLE_TENNIS = SportId("TABLE_TENNIS")
        val CRICKET = SportId("CRICKET")
    }
}

class Sport : JSONMappable {

    val id: SportId
    val name: String
    val isRacket: Boolean
    val properties: List<Property>
    val warnings: List<SportWarning>

    constructor(id: SportId, name: String, isRacket: Boolean, properties: List<Property>, warnings: List<SportWarning>) {
        this.id = id
        this.name = name
        this.isRacket = isRacket
        this.properties = properties
        this.warnings = warnings
    }

    @Throws(JSONException::class)
    constructor(json: JSONObject) : super(json) {
        id = SportId(json.getAny("sport_id"))
        name = json.getString("name")
        isRacket = json.getBoolean("is_racket")
        properties = json.getJSONArray("properties").flatMap { f ->
            try {
                Property(f)
            } catch (t: Exception) {
                null
            }
        }
        warnings = json.optJSONArray("warnings")?.flatMap {
            try {
                SportWarning(it)
            } catch (t: Exception) {
                null
            }
        } ?: listOf()
    }

    fun propertyWithId(id: PropertyId) =
        properties.firstOrNull { it.id == id }

    fun warningFor(propertyId: PropertyId, optionId: PropertyOptionId) =
        warnings.firstOrNull { it.propertyId == propertyId && it.optionId == optionId }

    fun filterProperties(filterIds: List<PropertyId>): List<Property> = properties.filter { filterIds.map { id -> id.value }.contains(it.id.value) }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        return (other as? Sport)?.id == id
    }

    override fun hashCode(): Int = id.hashCode()
}
